### DISCUSSION TITLE
How do I bind byte[] / BitmapImage in Image Source?

### DISCUSSION THREAD

**Discussion by vitkuz573** [microsoft/microsoft-ui-xaml#8049] [Q&A] [answered]
An exception is thrown when trying to bind `byte[]` in Source:

```
"System.InvalidCastException" in WinRT.Runtime.dll
WinRT information: No such interface supported

Exception thrown: "System.Runtime.InteropServices.COMException" in WinRT.Runtime.dll
WinRT information: No such interface supported

Exception thrown: "System.Runtime.InteropServices.COMException" in System.Private.CoreLib.dll
WinRT information: No such interface supported
```

Tried to bind `BitmapImage`:

```csharp
    public static async Task<BitmapImage> ImageFromBytes(this byte[] bytes)
    {
        var image = new BitmapImage();
 
        using (var stream = new InMemoryRandomAccessStream())
        {
            await stream.WriteAsync(bytes.AsBuffer());
            stream.Seek(0);
            await image.SetSourceAsync(stream);
        }
 
        return image;
    }
```

And a few other options, but another exception is thrown already:

```
Exception thrown: "System.Runtime.InteropServices.COMException" in System.Private.CoreLib.dll
Exception thrown: "System.AggregateException" in System.Private.CoreLib.dll
Exception thrown: "System.AggregateException" in System.Private.CoreLib.dll
```

New to WinUI, I don't understand why it doesn't work

**Comment by castorix** [comment, marked as answer]
Instead of BitmapImage I generally use WriteableBitmap, like for example in the _GetCaptureWriteableBitmap_ function in
[CMediaEngine.cs](https://github.com/castorix/WinUI3_MediaEngine/blob/master/CMediaEngine.cs)
(then Image.Source = WriteableBitmap to display it in an Image control)

**Reply by vitkuz573** [reply]
Sorry, but for some reason I get a `System.Runtime.InteropServices.COMException` in `WinRT.Runtime.dll` and then in `System.Private.CoreLib.dll` on all attempts to get `WriteableBitmap`. Can I have a concrete code example where there is only a *raw array of bytes* and I need to output it to `Image`?

**Reply by castorix** [reply]
If I use a similar sample with LoadImage and a .bmp to get a hBitmap then GetDIBits to get an array of bytes, it still works (img1 is an Image control) :
		
		    IntPtr hBitmap = LoadImage(IntPtr.Zero, "E:\\Butterfly_blue.bmp", IMAGE_BITMAP, 0, 0, LR_LOADFROMFILE);
                BITMAP bm;
                GetObject(hBitmap, Marshal.SizeOf(typeof(BITMAP)), out bm);
                int nWidth = bm.bmWidth;
                int nHeight = bm.bmHeight;
                BITMAPV5HEADER bih = new BITMAPV5HEADER();
                bih.bV5Size = Marshal.SizeOf(typeof(BITMAPV5HEADER));
                bih.bV5Width = nWidth;
                bih.bV5Height = -nHeight;
                bih.bV5Planes = 1;
                bih.bV5BitCount = 32;
                bih.bV5Compression = BI_BITFIELDS;
                bih.bV5AlphaMask = unchecked((int)0xFF000000);
                bih.bV5RedMask = 0x00FF0000;
                bih.bV5GreenMask = 0x0000FF00;
                bih.bV5BlueMask = 0x000000FF;

                IntPtr hDC = CreateCompatibleDC(IntPtr.Zero);
                IntPtr hBitmapOld = SelectObject(hDC, hBitmap);
                int nNumBytes = (int)(nWidth * 4 * nHeight);
                byte[] pBytes = new byte[nNumBytes];
                int nScanLines = GetDIBits(hDC, hBitmap, 0, (uint)nHeight, pBytes, ref bih, DIB_RGB_COLORS);

                WriteableBitmap wb = new WriteableBitmap(nWidth, nHeight);
                await wb.PixelBuffer.AsStream().WriteAsync(pBytes, 0, pBytes.Length);               
                img1.Source = wb;

                SelectObject(hDC, hBitmapOld);
                DeleteDC(hDC);

with 

        private const int IMAGE_BITMAP = 0;
        private const int IMAGE_ICON = 1;
        private const int IMAGE_CURSOR = 2;
        private const int IMAGE_ENHMETAFILE = 3;

        private const int LR_DEFAULTCOLOR = 0x00000000;
        private const int LR_MONOCHROME = 0x00000001;
        private const int LR_COLOR = 0x00000002;
        private const int LR_COPYRETURNORG = 0x00000004;
        private const int LR_COPYDELETEORG = 0x00000008;
        private const int LR_LOADFROMFILE = 0x00000010;

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Auto)]
        private static extern IntPtr LoadImage(IntPtr hInst, string lpszName, UInt32 uType, int cxDesired, int cyDesired, UInt32 fuLoad);

**Reply by vitkuz573** [reply]
I understand that you have an image from which you get an array of bytes and then work with it, but my situation is different: from an application on a remote PC I get a screenshot of its screen as `byte[]`, and as I understand I need to use only the following lines of this code:

```csharp
            WriteableBitmap wb = new WriteableBitmap(nWidth, nHeight);
            await wb.PixelBuffer.AsStream().WriteAsync(pBytes, 0, pBytes.Length);               
            img1.Source = wb;
```

But in this case I get a `COMException`

**Reply by vitkuz573** [reply]
Here's an example: code from viewmodel:

```csharp
var bytes = dto.ImageBytes.ToByteArray();

var wb = new WriteableBitmap(1920, 1080);
await wb.PixelBuffer.AsStream().WriteAsync(bytes, 0, bytes.Length);

CurrentScreenshot = wb; // This line causes a `COMException` to be thrown (UPD: not sure)
```

UPD: `COMException` is thrown even if you just create a `WriteableBitmap` instance

**Reply by castorix** [reply]
If I fill manually the byte array for example with 1 red pixel, it is ok too :

                byte[] pBytes = new byte[4];
                pBytes[0] = 0;
                pBytes[1] = 0;
                pBytes[2] = 255;
                pBytes[3] = 0;
                WriteableBitmap wb = new WriteableBitmap(1, 1);
                await wb.PixelBuffer.AsStream().WriteAsync(pBytes, 0, pBytes.Length);               
                img1.Source = wb;

**Reply by vitkuz573** [reply]
This only seems to happen when I do it in the ViewModel code, but if I write this code in the *.xaml.cs file, it works. I don't get it. I'll look into it, thanks!

**Reply by DarranRowe** [reply]
What is the type of CurrentScreenshot? For that kind of assignment to work, it must be along the inheritance hierarchy of [WritableBitmap](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.imaging.writeablebitmap?view=windows-app-sdk-1.2).

**Reply by vitkuz573** [reply]
> What is the type of CurrentScreenshot?

Is an observable property (CommunityToolkit MVVM)

```csharp
[ObservableProperty]
private WriteableBitmap currentScreenshot;
```

The thing is that even just creating an instance of `WriteableBitmap` throws `COMException`. And this happens only when writing code in ViewModel, but if you do the same in the file `*.xaml.cs`, then everything works fine

**Comment by DarranRowe** [comment]
Are the bytes raw pixel bytes or are they the contents of a bitmap file?
The source is assumed to be a bitmap file which has a bitmap header and other metadata, so that is probably why it is failing.
There are several options for getting raw bitmap pixels into an image source though.

1) [SoftwareBitmapSource](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.imaging.softwarebitmapsource?view=windows-app-sdk-1.2) allows you to use a [SoftwareBitmap](https://learn.microsoft.com/en-us/uwp/api/windows.graphics.imaging.softwarebitmap?view=winrt-22621) as a source. You can load this using bitmap bits.

2) [WritableBitmap](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.imaging.writeablebitmap?view=windows-app-sdk-1.2) allows you to set the size of the bitmap and then access the pixel buffer.

3) [SurfaceImageSource](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.imaging.surfaceimagesource?view=windows-app-sdk-1.2) allows you to use interop with Direct3D/Direct2D to write to the surface.

**Reply by vitkuz573** [reply]
Yes, it is a raw array of bytes
